### PR TITLE
Merak-topo:  create multilayer topology and handle return msg

### PR DESCRIPTION
**Pull request by zzengcs**
fix the conflict and create new pr for creating multilayer topo

### CODE REVIEW

**Comment by cj-chung** [comment]
@zzengcs Please give description for what  features or fix you did in this PR.

File: services/merak-topo/database/entities.go [hunk 1]
```diff
@@ -26,37 +26,44 @@ const (
...
-	Name string `json:"name"`
-	Intf string `json:"nic"`
-	Ip   string `json:"ip"`
-}
+// type Vport struct {
+// 	Id   string `json:"id"`
+// 	Name string `json:"name"`
+// 	Intf string `json:"nic"`
+// 	Ip   string `json:"ip"`
+// }
```
Comment (cj-chung) [inline services/merak-topo/database/entities.go:50]: @zzengcs If you don't need it, please remove it from the code.

File: services/merak-topo/database/entities.go [hunk 2]
```diff
@@ -26,37 +26,44 @@ const (
...
+	Peer_intf  string        `json:"peer_intf"`
+	Peer_ip    string        `json:"peer_ip"`
+	Status     ServiceStatus `json:"status"`
 }
 
 type TopologyData struct {
 	Topology_id string  `json:"topology_id"`
 	Vnodes      []Vnode `json:"vnodes"`
-	Vlinks      []Vlink `json:"vlinks"`
+	// Vlinks      []Vlink `json:"vlinks"`
```
Comment (cj-chung) [inline services/merak-topo/database/entities.go:68]: @zzengcs Please remove the line you've commented out.

File: services/merak-topo/grpc/service/service.go [hunk 3]
```diff
@@ -106,7 +109,9 @@ func (s *Server) TopologyHandler(ctx context.Context, in *pb.InternalTopologyInf
 			//
 		default:
 			// pb.TopologyType_TREE
-			err_create := handler.Create(k8client, topo_id, uint32(aca_num), uint32(rack_num), uint32(aca_per_rack), uint32(cgw_num), data_plane_cidr, &returnMessage)
+			// err_create := handler.Create(k8client, topo_id, uint32(aca_num), uint32(rack_num), uint32(aca_per_rack), uint32(cgw_num), data_plane_cidr, &returnMessage)
```
Comment (cj-chung) [inline services/merak-topo/grpc/service/service.go:112]: @zzengcs Remove the commented out line.

File: services/merak-topo/handler/handler.go [hunk 4]
```diff
@@ -156,28 +107,192 @@ func Create(k8client *kubernetes.Clientset, topo_id string, aca_num uint32, rack
...
+	if err_deploy != nil {
+		return fmt.Errorf("topology deployment error %s", err_deploy)
+	}
+
 	return nil
+
 }
 
+// func Create(k8client *kubernetes.Clientset, topo_id string, aca_num uint32, rack_num uint32, aca_per_rack uint32, cgw_num uint32, data_plane_cidr string, returnMessage *pb.ReturnTopologyMessage) error {
+
```
Comment (cj-chung) [inline services/merak-topo/handler/handler.go:152]: @zzengcs Please remove the code you've commented out.

File: services/merak-topo/handler/topo_deploy.go [hunk 5]
```diff
@@ -38,11 +38,11 @@ import (
 )
 
 var (
-	ACA_IMAGE = "meraksim/merak-agent:test2"
-	OVS_IMAGE = "yanmo96/ovs_only:latest"
-	RYU_IP    = "ryu.merak.svc.cluster.local"
-	RYU_PORT  = "6653"
-	Ctx       = context.Background()
+	// ACA_IMAGE = "meraksim/merak-agent:311f5f6c"
+	// OVS_IMAGE = "yanmo96/ovs_only:latest"
```
Comment (cj-chung) [inline services/merak-topo/handler/topo_deploy.go:42]: @zzengcs Remove two commented-out lines here.

File: services/merak-topo/handler/topo_deploy.go [hunk 6]
```diff
@@ -52,12 +52,13 @@ var (
...
 
-func CreateTopologyClasses(client dynamic.Interface, name string, links []map[string]interface{}) error {
+func CreateTopologyClasses(client dynamic.Interface, name string, links []database.Vlink) error {
 	rc := NewTopologyClass(name, links)
 
 	_, err := client.Resource(topologyClassGVR).Namespace(namespace).Create(Ctx, rc, metav1.CreateOptions{})
 
-	log.Printf("Creating TopologyClass %s", name)
+	// log.Printf("Creating TopologyClass %s", name)
+	// log.Printf("topology class details %s", rc)
```
Comment (cj-chung) [inline services/merak-topo/handler/topo_deploy.go:61]: @zzengcs Remove the commented-out lines.

File: services/merak-topo/handler/topo_deploy.go [hunk 7]
```diff
@@ -69,15 +70,15 @@ func CreateTopologyClasses(client dynamic.Interface, name string, links []map[st
...
+	_, err := client.Resource(topologyClassGVR).Namespace(namespace).Get(Ctx, name, metav1.GetOptions{})
 
 	log.Printf("Get TopologyClass %s", name)
 
 	if err != nil {
 		return fmt.Errorf("failed to create topologyClass %s", err)
 	}
 
-	fmt.Printf("Get %v topology data: %v", name, data)
+	// fmt.Printf("Get %v topology data: %v", name, data)
```
Comment (cj-chung) [inline services/merak-topo/handler/topo_deploy.go:81]: @zzengcs Remove the line.

**Comment by zzengcs** [comment]
> @zzengcs Please give description for what features or fix you did in this PR.

This PR contains the feature of merak-topo for creating multi-layer topology structure with the given inputs from scenario manager on the number of vhost, the number of racks, the number of ports per vswitch, and the image of pods.

File: services/merak-topo/handler/topo_gen.go [hunk 8]
```diff
@@ -27,269 +29,424 @@ func GenUUID() string {
...
 		//
 	default: //16
-		for i := 0; i < upper; i++ {
-			ip = strings.Split(data_plane_cidr, ".")[0] + "." + strings.Split(data_plane_cidr, ".")[1] + "." + strconv.FormatInt(int64(k1), 10) + "." + strconv.FormatInt(int64(i+1), 10) + "/16"
-			ips_gen = append(ips_gen, ip)
-		}
+		i := (vhost_idx) / upper
+		// if (vhost_idx)%upper > 0 {
+		// 	i = i + 1
+		// }
```
Comment (cj-chung) [inline services/merak-topo/handler/topo_gen.go:43]: @zzengcs Remove them.

File: services/merak-topo/handler/topo_gen.go [hunk 9]
```diff
@@ -27,269 +29,424 @@ func GenUUID() string {
...
-			ips = append(ips, ip_gen_sub(count, iter, data_plane_cidr)...)
+	j := init_idx_vs
+	k := ports_per_vswitch
+	for i := 0; i < num_of_vs; i++ {
+		lower_bound := i * k
+		var upper_bound int
+		if i*k+k-1 < len(racks) {
+			upper_bound = lower_bound + k
+		} else {
+			upper_bound = len(racks)
```
Comment (cj-chung) [inline services/merak-topo/handler/topo_gen.go:67]: @zzengcs if possible, we should briefly explain what algorithm you're using here.

File: services/merak-topo/database/entities.go (hunk 1 above)
Comment (zzengcs) [reply services/merak-topo/database/entities.go]: addressed

File: services/merak-topo/database/entities.go (hunk 2 above)
Comment (zzengcs) [reply services/merak-topo/database/entities.go]: addressed

File: services/merak-topo/grpc/service/service.go (hunk 3 above)
Comment (zzengcs) [reply services/merak-topo/grpc/service/service.go]: addressed

File: services/merak-topo/handler/topo_deploy.go (hunk 5 above)
Comment (zzengcs) [reply services/merak-topo/handler/topo_deploy.go]: addressed

File: services/merak-topo/handler/topo_deploy.go (hunk 6 above)
Comment (zzengcs) [reply services/merak-topo/handler/topo_deploy.go]: addressed

File: services/merak-topo/handler/topo_deploy.go (hunk 7 above)
Comment (zzengcs) [reply services/merak-topo/handler/topo_deploy.go]: addressed

File: services/merak-topo/handler/topo_gen.go (hunk 8 above)
Comment (zzengcs) [reply services/merak-topo/handler/topo_gen.go]: addressed

**Comment by zzengcs** [comment]
Separate this PR into two PR, One is covered by the PR #117 , another will be submitted later for Return Message

**Pull request closed without merging**